src/App.js

```diff
@@ -37,77 +128,429 @@ function App(props) {
+        });
+    }
+
+    // Edit item
+    const onEditItem = (id, text, completed, priority) => {
+        setUndoStack(undoStack.concat([{
+            type: UndoType.Item,
+            op: UndoOp.Edit,
+            oldList: subcollectionName,
+            newList: null,
+            data: itemsData
+        }]));
+
+        updateDoc(doc(db, collectionPath + subcollectionName, id), {
+           text: text,
+           completed: completed,
+           priority: priority,
+        });
+    }
+
+    // Delete item by id
+    const onDeleteItem = (id) => {
+        setUndoStack(undoStack.concat([{
+            type: UndoType.Item,
+            op: UndoOp.Delete,
+            oldList: subcollectionName,
+            newList: null,
+            data: itemsData
+        }]));
+
+        deleteDoc(doc(db, collectionPath + subcollectionName, id));
     }
 
     // Remove completed items
     const onRemoveCompleted = () => {
-        setData(data.filter((item) => !item.completed));
+        setUndoStack(undoStack.concat([{
+            type: UndoType.Item,
+            op: UndoOp.Delete,
+            oldList: subcollectionName,
+            newList: null,
+            data: itemsData
+        }]));
+
+        const batch = writeBatch(db);
+
+        itemsData.forEach((item) => {
+            if (item.completed) {
+                batch.delete(doc(db, collectionPath + subcollectionName, item.id));
+            }
+        });
+
+        batch.commit();
     }
 
-    // Add item
-    const onAddItem = (text) => {
-        setData(data.concat([{id: counter, text: text, completed: false}]));
-        setCounter(counter + 1);
+    // Add list
+    const onAddList = (name) => {
+        setUndoStack(undoStack.concat([{
+            type: UndoType.List,
+            op: UndoOp.Add,
+            oldList: null,
+            newList: name,
+            data: [],
+        }]));
+
+        const newLists = [...userData.lists];
+        newLists.push(name);
+        updateDoc(doc(db, collectionName, userDocumentName), {
+            lists: newLists
+        });
+
+        // Update state to be on newly added list
+        setSubcollectionName(name);
     }
 
-    // Delete item by id
-    const onDeleteItem = (id) => {
-        setData(data.filter((item) => item.id !== id));
+    // Edit list name
+    const onEditList = (oldName, newName) => {
+        setUndoStack(undoStack.concat([{
+            type: UndoType.List,
+            op: UndoOp.Edit,
+            oldList: oldName,
+            newList: newName,
+            data: itemsData,
+        }]));
+
+        // Firestore does not support renaming collections
+        // To rename list, must delete old list and re-add all of its items under new collection name
+        // Use batch to make this operation atomic
+        const batch = writeBatch(db);
+
+        // Remove list from user lists array and replace with new list containing new list name
+        const newLists = userData.lists.map((list) => {
```

Comment:
Simpler approach would be to have your docId be a unique ID and then have the list name stored in the document. That way, one simple change to edit a list name.

Reply:
I looked online and there wasn't an easy way to get the subcollections of a document. So, we went this way of having an array of the list names for each user document. So, if we want to rename a list within the array, we would have to replace the array with the name substituted.

Reply:
I think you'd find it much simpler if you had a top-level lists collection, and then had each list have an ID (stored in the list as well as the ID of the document), along with the name.  Then, you could have the tasks of the list be a subcollection of the list.